### PR TITLE
[DDW-500] Implement native tokens

### PR DESCRIPTION
**Pull request by DeeJayElly**
This PR implements design for native tokens.

Design:
https://zpl.io/VDeA6BJ - multi-send
https://zpl.io/VxEwEPR - multi-send confirmation dialog
https://zpl.io/bAr9Roe - wallet with tokens
https://zpl.io/2vv8qWe - wallet with tokens - popover
https://zpl.io/VQlz9XP - tx details

## Issues 

Issue list is in this [document](https://docs.google.com/document/d/1R5lhnDGMIT6JguzGc0fiYTPDv4QiNTm4hSjMFIyab8w/edit?usp=sharing)

## Scenarios

1. User opens up Send token page and sees receiver input field
2. User starts typing/pasting address in receiver input field. 
3. If address is valid  input field is rendered bellow which has ADA as currency. 
4. User can select other tokens by clicking on + Add another token button
5. Input field with dropdown to choose tokens is presented
6. All the tokens in the dropdown are ordered alphabetically after default ADA.
7. When you click on + Add another token button new token input field will be added at the bottom
8. Every time user selects some token from the dropdown, that token is removed from the list for future selection on other assets.
9. When there is no more tokens available in the dropdown, + Add another token button will be disabled
10. ADA input field will be always render at the top of the list, first item and it cant be removed (ADA is mandatory as deposit when sending a transaction with tokens)
11. For ADA input field amount will always have format with 6 decimal places, while for tokens formatting will be with round numbers (without fractions).

## Todos

- [x] [Old TODO list](https://docs.google.com/document/d/1itdqdyxZcySF__PXdRTsKbX1yR00JNHeeFLvxjcVzuk/edit)

### New ones

- [x] Add proper Confirmation dialog [@daniloprates ]
- [x] Hide WalletSummaryAssets when syncing wallets [@nikolaglumac ]
- [x] Reset validation when reseting Send form [@DeeJayElly ]
- [x] On reset Send form, validation should reset errors [@DeeJayElly ]
- [x] Fix vertical alignment of "No recent Tx" message [@nikolaglumac ]
- [x] Estimate fee when there is token qty but not ADA qty [@DeeJayElly ]
- [x] Fix minimum ADA required (Either min UTXO or fee calculation) [@DeeJayElly ]
- [x] Fix "Add Token" isDisabled state handling (after adding and removing tokens) [@DeeJayElly ]
- [x] If there's 0 decimal places, no dot should be allowed [@DeeJayElly ]
- [x] Token input placeholder to follow the decimal figures [@DeeJayElly ]
- [x] We should focus the token amount input field after clicking on the "+ add new token" button [@DeeJayElly ]
- [x] Fix PopOver disappearing when hovering (also affects the stake pool tooltip on the Delegation center) [@daniloprates ]
- [x] Improve copy button position (should be inline with the text regardless of the text length) [@daniloprates ]
- [x] In the Tx filter, remove the word "Ada" in "Ada sent" and "Ada received" [@daniloprates ]
- [ ] ~Show AssetToken popOver in the Token input [@daniloprates ]~ (_Not possible, for now. The hover element needs to be the hidden Dropdown input. If the PopOver hovering element gets on top, the PopOver won't work. We can improve this in the future, but for now it's too risky_)
- [x] Hide token when no fingerprint [@nikolaglumac ]
- [x] Sort Tokens in the Summary by Fingerprint [@nikolaglumac ]
- [x] Update [copy](https://docs.google.com/document/d/13QUArUgQuK4ZwePRuDHyZkWbiJnUEUMggI9QhwWsNO0/edit) of Wallet Receive screen [@nikolaglumac ]
- [x] Check if the fingerprint is prefixed with 'token' or 'asset' [@nikolaglumac ]
- [x] Investigate `NaN` error upon entering typing comma character in Numeric Input [@nikolaglumac ]

### Tomorrow (Nice to have)

- Export CSV update 

### @DeeJayElly 

- [x] Summary screen - loading state for the asset section
- [x] Summary screen - always show ADA as the first asset in the assets section
- [x] Receive screen - keep left and right paddings aligned with the existing screens (should be reduced)
- [x] Assets should be renamed to Tokens everywhere
- [x] Finalize multi-send dialog
- [x] Finalize multi-send confirmation dialog
- [x] Add translations

###  @daniloprates 
- [x] Send Minimum ada required to Send form
- [x] Remove Input from Tx (Still needs double check)
- [x] Add AssetToken to Add to Transaction and change it to the [correct layout](https://app.zeplin.io/project/57acfaa7436a750316161a8b/screen/602d29d0b5e9c02345d7a634)
- [x] Convert all AssetToken quantities into BigNumber
- [x] Fix the Tx in the Wallets Summary screen
- [x] Wallet Dropdown styling issues

### @nikolaglumac 
- [x] Parse transaction total tokens moved
- [x] Set listAssets polling (60 sec interval)
- [x] Re-check Summary screen with repeated Asset data

## Pending issues

- [x] Paddings - left/right - reduce it to 20px on the form container
- [x] Default token in the dropdown is always ADA - first token
- [x] Fractions should be taken out from the metadata and default is 0 if using tokens, only show decimal places for ADA
- [x] Tokens should be order alphabetically
- [x] New receiver should never render token field before start typing in the field and passing validation
- [x] We shouldn’t accept same address - show tooltip error when trying to add same address as receiver
- [x] Ada field will be rendered always on bottom, you can’t change its currency but you can change its value. Every time you select a new token to add it, we will remove it from the dropdown list (reduce items in the list)
- [x] When there are no more tokens in the dropdown, Add new token button will be disabled

## UI issues document:

https://docs.google.com/document/d/1R5lhnDGMIT6JguzGc0fiYTPDv4QiNTm4hSjMFIyab8w/edit

## Screenshots

Screen Shot 2021-01-19 at 10 49 44 AM

Screen Shot 2021-01-19 at 10 49 59 AM

Screen Shot 2021-01-29 at 8 33 04 PM

Screen Shot 2021-01-29 at 8 31 11 PM

Screen Shot 2021-01-29 at 8 32 29 PM

Screen Shot 2021-01-29 at 8 30 17 PM

Screen Shot 2021-01-29 at 8 31 19 PM

Screen Shot 2021-01-29 at 8 30 04 PM

Screen Shot 2021-01-25 at 4 01 05 AM

Screen Shot 2021-01-25 at 4 01 22 AM

Screen Shot 2021-01-25 at 8 10 40 AM

Screen Shot 2021-01-25 at 8 10 29 AM

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1614435277045400)
- [ ] Test

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 1]
```diff
@@ -108,7 +108,6 @@ type Props = {
     amount: number
   ) => Promise<BigNumber>,
   currentNumberFormat: string,
-  walletAmount: BigNumber,
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:111]: @DeeJayElly please rever this change.

File: storybook/stories/_support/utils.js [hunk 2]
```diff
@@ -71,6 +71,36 @@ export const generateWallet = (
     delegatedStakePoolId: get(delegatedStakePool, 'id'),
   });
 
+export const generateNativeTokenWallet = (
```
Comment (nikolaglumac) [inline storybook/stories/_support/utils.js:74]: @DeeJayElly please change this name 🙏

File: source/renderer/app/utils/transaction.js [hunk 3]
```diff
@@ -76,6 +76,7 @@ export const isTransactionAmountInFilterRange = (
   transaction: WalletTransaction
 ) => {
   const { amount } = transaction;
+  const amountAbsoluteValue = amount ? amount.absoluteValue() : null;
```
Comment (nikolaglumac) [inline source/renderer/app/utils/transaction.js:79]: @DeeJayElly please improve this. All amounts **need** to be in BigNumber format.

**Review by nikolaglumac — CHANGES_REQUESTED** [review]

File: source/renderer/app/actions/wallets-actions.js [hunk 4]
```diff
@@ -60,6 +60,7 @@ export default class WalletsActions {
   chooseWalletExportType: Action<{
     walletExportType: WalletExportTypeChoices,
   }> = new Action();
+  setNativeTokenWallet: Action<{ walletId: string }> = new Action();
```
[the remark this answers is not in the document]
Comment (DeeJayElly) [reply source/renderer/app/actions/wallets-actions.js]: Fixed

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 1 above)
Comment (DeeJayElly) [reply source/renderer/app/components/wallet/WalletSendForm.js]: Fixed

File: storybook/stories/_support/utils.js (hunk 2 above)
Comment (DeeJayElly) [reply storybook/stories/_support/utils.js]: Fixed

File: source/renderer/app/utils/transaction.js (hunk 3 above)
Comment (DeeJayElly) [reply source/renderer/app/utils/transaction.js]: Fixed

**Comment by nikolaglumac** [comment]
@DeeJayElly please merge latest develop...

File: source/renderer/app/api/api.js [hunk 5]
```diff
@@ -583,6 +593,46 @@ export default class AdaApi {
...
+      return new Promise((resolve) =>
+        resolve({ assets, total: response.length })
+      );
+    } catch (error) {
+      logger.error('AdaApi::getAssets error', { error });
+      throw new ApiError(error);
+    }
+  };
+
+  getAsset = async (request: GetAssetRequest): Promise<GetAssetResponse> => {
```
Comment (nikolaglumac) [inline source/renderer/app/api/api.js:616]: @DeeJayElly @daniloprates this one should never be used and it should be removed from the code.

**Comment by nikolaglumac** [comment]
@alexander-rukin please test this one and update the report document (please add the link to it to the PR description).

File: source/renderer/app/components/layout/VerticalFlexContainer.scss [hunk 6]
```diff
@@ -2,6 +2,5 @@
   display: flex;
   flex-direction: column;
   height: 100%;
-  overflow-y: overlay;
```
Comment (nikolaglumac) [inline source/renderer/app/components/layout/VerticalFlexContainer.scss:5]: @DeeJayElly are you sure this change is required and that it doesn't break anything?

**Review by darko-mijic — APPROVED** [review]

**Review by alan-mcnicholas — APPROVED** [review]
Incredible work as always!

**Review by miorsufianiohk — APPROVED** [review]
LGTM 👍 . Tested on 16862.

**Pull request merged**